### PR TITLE
Add harvester count and correct remote plots calculation

### PR DESCRIPTION
**Pull request by sashacmc**
Add connected harvester count to farm summary.
And correct remote plots calculation
(it requires change self_hostname to '0.0.0.0' on remote harvester hosts)

### CODE REVIEW

**Comment by fiveangle** [comment]
This has been working excellently in my small testbed running this over 1.1.2, moving to prod now.  Thanks Alex !

**Comment by sashacmc** [comment]
Dear @fiveangle, 
I fixed lint warnings, but I have no idea why some test failed, seems it is not connected with my change.
Please review it.

**Comment by fiveangle** [comment]
Those rate limit errors do seem entirely unrelated.  I didn't dig too far, but that they seem to trickle down from mempool, I suspect it may be related to the transaction caching work being done. I've not seen any such errors on my running system.
 
In thinking about rounding out this PR, I was thinking it would be nice to be able to report the response times of each harvester in the `farming_info from peer` log entry.  I was trying to think about how one could save the timestamp of the `new_signage_point_harvester to peer` for each querying harvester, then report the time delta between that timestamp and the timestamp that the `farming_info from peer` for that peer comes back, and simply report it on that line.  Do we know what data the peer sends back for that response ?  I'd imagine it's just a go/no-go for the proof, at which point just adding the farmer-side-calculated time delta would be all we could hope for without changing the harvester side.  Regardless, one of the big issues plaguing many today is unbeknownst to them high lookup times (for whatever reason) so logging the overall round-trip time for each harvester would be an indispensable troubleshooting tool.  I see 1.1.3 is slated to warn on >5s lookup, but that is surely local at the harvester, which leaves lots of places to watch.  Let me know if you have any thoughts on this !
 
 ```
2021-05-01T11:52:08.406 farmer farmer_server              : INFO     -> new_signage_point_harvester to peer 10.0.111.248 490b14840c1d377ceec2b9ee9a13d1926eda2f7ee386602ec83d52f870f07d57
2021-05-01T11:52:08.407 farmer farmer_server              : INFO     -> new_signage_point_harvester to peer 10.0.111.12 3e138666d38402e2d8ae29d6169a63e2d4c04925ddb31fd92b14fa2be4d15009
2021-05-01T11:52:08.408 farmer farmer_server              : INFO     -> new_signage_point_harvester to peer 10.0.111.151 413e133151e4f92a16c9b9d496409a31b92d30774edad6bd61a7a47fa96a5900
2021-05-01T11:52:08.408 farmer farmer_server              : INFO     -> new_signage_point_harvester to peer 10.0.111.16 cd8b41b44ffce421100adb57b698d6fe79368567c294a38aabbd76f5289eeb77
2021-05-01T11:52:08.408 farmer farmer_server              : INFO     -> new_signage_point_harvester to peer 10.0.111.121 3c8f8e5a461dab32bb1ee1736e6cb6f1fe00775133e12e210f3fe76b0cfa6161
2021-05-01T11:52:08.409 farmer farmer_server              : INFO     <- farming_info from peer 490b14840c1d377ceec2b9ee9a13d1926eda2f7ee386602ec83d52f870f07d57 10.0.111.248
2021-05-01T11:52:08.411 farmer farmer_server              : INFO     <- farming_info from peer 3e138666d38402e2d8ae29d6169a63e2d4c04925ddb31fd92b14fa2be4d15009 10.0.111.12
2021-05-01T11:52:08.417 farmer farmer_server              : INFO     <- farming_info from peer cd8b41b44ffce421100adb57b698d6fe79368567c294a38aabbd76f5289eeb77 10.0.111.16
2021-05-01T11:52:08.552 farmer farmer_server              : INFO     <- farming_info from peer 413e133151e4f92a16c9b9d496409a31b92d30774edad6bd61a7a47fa96a5900 10.0.111.151
2021-05-01T11:52:09.318 farmer farmer_server              : INFO     <- farming_info from peer 3c8f8e5a461dab32bb1ee1736e6cb6f1fe00775133e12e210f3fe76b0cfa6161 10.0.111.121
```

**Comment by sashacmc** [comment]
Dear @fiveangle, as part of this PR make sense to measure only the get_plots request time, But it is other time and will not help with diagnostic.
I can implement the time delta measure between new_signage_point_harvester and farming_info as separate PR.

**Comment by fiveangle** [comment]
> Dear @fiveangle, as part of this PR make sense to measure only the get_plots request time, But it is other time and will not help with diagnostic.
> I can implement the time delta measure between new_signage_point_harvester and farming_info as separate PR.

@sashacmc please check your repository as I have submitted a PR to normalize the output to match the current local harvester API:

image

**Comment by kridigger** [comment]
@mariano54 what do you think about this PR? For me it also works very well and I'd really like to see it in the next release :)

**Comment by kridigger** [comment]
@sashacmc it seems you must request a review from somebody with write access? in the top right you should be able to click in the field "reviewers"

**Comment by sashacmc** [comment]
> @sashacmc it seems you must request a review from somebody with write access? in the top right you should be able to click in the field "reviewers"

Unfortunately, this field is inactive.

**Comment by kridigger** [comment]
oh that's unfortunate! so a mention must suffice!

**Review by lamba09 — COMMENTED** [review]
I like this feature. The code looks good to me, what do you think @mariano54 ?

**Comment by losingle** [comment]
> Add connected harvester count to farm summary.
> And correct remote plots calculation
> (it requires change self_hostname to '0.0.0.0' on remote harvester hosts)

You must have access to port 8560 open on the firewall on the remote harvester

**Comment by kridigger** [comment]
@wjblanke could you perhaps have a look at this PR? Thank you! :smiling_face_with_three_hearts:

**Review by b1u3h4t — APPROVED** [review]
It is a good feature, hope to merge.

**Review by b1u3h4t — COMMENTED** [review]
It is good feature.

**Comment by fiveangle** [comment]
i guess because we got 2 pr subs for the same great and critical feature, we got neither of them 🤣

**Comment by sashacmc** [comment]
I already lost hope of seeing it in the main, so I just apply it as a patch when updating

**Comment by fiveangle** [comment]
> I already lost hope of seeing it in the main, so I just apply it as a patch when updating

yeah, me too... it's just a little trickier for my windows system, which I was hoping to prevent, but I've got the workflow down to apply the patch.  still a PITA tho :)

Thanks again @sashacmc for putting this together. It has been instrumental for alerting me to harvester problems across my hodge-podge of old repurposed equipment (which seems to love to fail randomly ; )

**Comment by kridigger** [comment]
perhaps @hoffmang9 can help pushing this PR over the finish line? :)

**Comment by savager** [comment]
Hi sorry, I am hoping for a little clarification here. I patched my environment with this and I get the following connection error to my harvesters.

`Connection error. Check if harvester is running at 10.0.1.178:8560 Cannot connect to host 10.0.1.178:8560 ssl:<ssl.SSLContext object at 0x7fcd7a644b40> [Connect call failed ('10.0.1.178', 8560)]
Connection error. Check if harvester is running at 10.0.1.176:8560 Cannot connect to host 10.0.1.176:8560 ssl:<ssl.SSLContext object at 0x7fcd7a644240> [Connect call failed ('10.0.1.176', 8560)]
Farming status: Farming
Total chia farmed: 0.0
User transaction fees: 0.0
Block rewards: 0.0
Last height farmed: 0
Harvester count: 3
Plot count: 20
Estimated network space: 10778.857 PiB`

I actually might be confused. I see in my farmer logs this error when starting my harvesters. Perhaps my harvesters are not connecting? That would be a major problem. 

`2021-05-25T13:29:59.895 farmer farmer_server              : INFO     Connection closed: 10.0.1.176, node id: XXXX
2021-05-25T13:29:59.895 farmer chia.farmer.farmer         : INFO     peer disconnected {'host': '10.0.1.176', 'port': 8448}`

I thought all was well because I do see the following logs on my harvesters:

`2021-05-25T13:53:27.071 harvester chia.harvester.harvester: INFO     0 plots were eligible for farming ee74332c06... Found 0 proofs. Time: 0.01751 s. Total XXX plots`

and then the corresponding logs on the farmer:

`2021-05-25T13:55:16.952 farmer farmer_server              : DEBUG    -> new_signage_point_harvester to peer 10.0.1.178 7f883d`

And they happen in tandem so I assumed all was working well. Appreciate any help to get this running, I think it's a nice feature and should be merged.

File: chia/cmds/farm_funcs.py [hunk 1]
```diff
@@ -12,24 +12,26 @@
...
 
 
-async def get_plots(harvester_rpc_port: int) -> Optional[Dict[str, Any]]:
+async def get_plots(harvester_rpc_port: int, harvester_rpc_hostname: str) -> Optional[Dict[str, Any]]:
     plots = None
     try:
         config = load_config(DEFAULT_ROOT_PATH, "config.yaml")
-        self_hostname = config["self_hostname"]
+        if harvester_rpc_hostname is None:
+            harvester_rpc_hostname = config["self_hostname"]
```
Comment (Jahorse) [inline chia/cmds/farm_funcs.py:25]: Could we add a check here to translate `localhost` into `0.0.0.0` so we don't need to modify our config to make this work?

**Comment by emlowe** [comment]
Thanks for the PR. We believe 1.2 resolves most of these issues. Please open a new PR for further enhancements

**Pull request closed without merging**